### PR TITLE
Propogate from master and Update Proxy Envvars for PodTemplates

### PR DESCRIPTION
**Pull request by waveywaves**

### CODE REVIEW

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java [hunk 1]
```diff
@@ -143,7 +154,7 @@ public static synchronized void addPodTemplate(PodTemplate podTemplate) {
     // clear out existing template with same name; k8s plugin maintains
     // list, not map
     removePodTemplate(podTemplate);
-
+    podTemplate = setProxyEnvVar(podTemplate);
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: return value is not needed here

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java [hunk 2]
```diff
@@ -159,6 +170,21 @@ public static synchronized void addPodTemplate(PodTemplate podTemplate) {
...
+    String httpProxy = System.getenv("http_proxy");
+    String httpsProxy = System.getenv("https_proxy");
+    if (httpProxy != null && !httpProxy.isEmpty()) {
+      envVars.add(new KeyValueEnvVar("http_proxy", httpProxy));
+    }
+    if (httpsProxy != null && !httpsProxy.isEmpty()) {
+      envVars.add(new KeyValueEnvVar("https_proxy", httpsProxy));
+    }
+    podTemplate.setEnvVars(envVars);
+    return podTemplate;
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: no need to return value

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java [hunk 3]
```diff
@@ -159,6 +170,21 @@ public static synchronized void addPodTemplate(PodTemplate podTemplate) {
     }
   }
 
+  protected static PodTemplate setProxyEnvVar(PodTemplate podTemplate) {
+    LOGGER.info("Adding Proxy EnvVar for Template "+podTemplate.getName());
+    List<TemplateEnvVar> envVars = new ArrayList<TemplateEnvVar>();
+    String httpProxy = System.getenv("http_proxy");
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: can you set these as constants:
private final static String HTTP_PROXY_ENV_VAR_NAME = "http_proxy";

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 2 above)
Comment (akram) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: no_proxy is not propagated !

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java [hunk 4]
```diff
@@ -208,12 +235,15 @@ protected static void addPodTemplate(BaseWatcher baseWatcher, String type, Strin
         String name = podTemplate.getName();
         // we allow configmap overrides of maven and nodejs, but not imagestream ones
         // as they are less specific/defined wrt podTemplate fields
+//        podTemplate = PodTemplateUtils.setProxyConfigToPodTemplate(podTemplate);
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: remove commented code

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java [hunk 5]
```diff
@@ -412,4 +442,31 @@ protected static boolean hasSlaveLabelOrAnnotation(Map<String, String> map) {
...
+
+
+  protected synchronized static void propogateProxyConfigToReservedPodTemplates(){
+    // Propogate Proxy config to node and maven podTemplates which cannot be changed otherwise
+    KubernetesCloud kubeCloud = JenkinsUtils.getKubernetesCloud();
+    LOGGER.info("Propogate proxy config to Reserved PodTemplates");
+    if (kubeCloud != null) {
+      List<PodTemplate> list = kubeCloud.getTemplates();
+      Iterator<PodTemplate> iter = list.iterator();
+      while (iter.hasNext()) {
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: Maybe more readable to use a `for( PodTemplate pt : list )`  loop

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java [hunk 6]
```diff
@@ -5,14 +5,22 @@
 import io.fabric8.kubernetes.api.model.ConfigMap;
 import io.fabric8.kubernetes.api.model.ObjectMeta;
 import io.fabric8.kubernetes.api.model.Pod;
-import io.fabric8.openshift.api.model.*;
+import io.fabric8.openshift.api.model.Image;
+import io.fabric8.openshift.api.model.ImageStream;
+import io.fabric8.openshift.api.model.ImageStreamStatus;
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: Can you optimize imports?
it looks like ImageStreamStatus is not used ? is it ?

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 5 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: This implementation is fine I think as we would be makin changes to the list while reading the list. Probably would get a concurrency based exception.

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 1 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: Here we arent using kubecloud api inside setProxyEnvVar to updat the variable. Makes more sense to have it I feel.

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 6 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: It is used in PodTemplateUtils itself

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 4 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: done

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 3 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: done

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 2 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: done

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 1 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: Understool :)

**Pull request merged**